Relax typing of classification instructions; opcode for dataref

rossberg opened the pull request:
Addresses #180 and #183.

rossberg commented:
@manoskouk, PTAL.

manoskouk commented on proposals/gc/MVP.md:
```diff
@@ -325,39 +325,44 @@ Tentatively, support a type of guaranteed unboxed scalars.
 * `br_on_func <labelidx>` branches if a reference is a function
-  - `br_on_func $l : [anyref] -> [anyref]`
-    - iff `$l : [funcref]`
+  - `br_on_func $l : [(ref null ht)] -> [(ref null ht)]`
```
I think the typing of all `br_on_*` instructions, including `br_on_cast`, could be further refined to preserve their argument's nullability, similarly to `ref.cast`.

rossberg replied:
Good point, PTAL!

manoskouk reviewed (commented):
LGTM with one comment.

manoskouk reviewed (approved):
LGTM, thanks!

rossberg commented:
Ah, yes. I included that change, and simplified their typing rules. While at it, I also removed the explicit type index on those instructions, which is no longer consistent with the rest of the language. PTYAL.

manoskouk commented:
Looks good, thanks!

manoskouk commented on proposals/gc/MVP.md:
```diff
@@ -379,29 +387,27 @@ TODO: Add the ability to generate new (non-canonical) RTT values to implement ca
+  - `br_on_cast $l : [t (rtt n? $t')] -> [t]`
+    - iff `$l : [t']`
+    - and `t <: dataref` or `t <: funcref`
+    - and `(ref $t) <: t'`
```
I think there are some aliased variables here. It should be:
`br_on_cast $l : [t1 (rtt n? $t2)] -> [t1]`
- iff `$l : [t3]`
- and `t1 <: (ref null data)` or `t1 <: funcref` (instead, `t1 <: anyref` if we go with the above suggestion)
- and `(ref $t2) <: t3`

rossberg replied:
There was a typo in the last line, which should have been `(ref $t')` (fixed). But I think the others are correct, considering that `$t'` and `t'` are different names and in different categories.

rossberg commented on proposals/gc/MVP.md:
```diff
@@ -379,29 +387,27 @@ TODO: Add the ability to generate new (non-canonical) RTT values to implement ca
-    - and `(type $t) <: ht`
+* `ref.test` tests whether a reference value's [runtime type](#values) is a [runtime subtype](#runtime) of a given RTT
+  - `ref.test : [t' (rtt n? $t)] -> [i32]`
+    - iff `t' <: dataref` or `t' <: funcref`
```
Ah, no, the purpose of introducing the separate classification instructions actually was to simplify the test/cast instructions, so that they only operate on references that are known to have an RTT. Otherwise there function would overlap with the classification instructions. See #150 for more context.

You are right about the null thing, fixed.

rossberg reviewed (commented):
Thanks, I pushed a quick fix to the typos you point out.
